### PR TITLE
Make tracer_compound a required field

### LINKED ISSUE #312: PeakData results cause a `NoReverseMatch` Exception when an animal has no tracer specified
### Problem

Viewing the PeakData for studies where some animals have no tracer specified causes a `NoReveserMatch` Excpetion.

The animal `MN_exp027e_M1` was loaded incorrectly without a tracer. It appears that the tracer should be `arginine`. It's possible (likely) the other animals in the study were also loaded incorrectly. However, we have determined that TraceBase *should* support Animals with no tracer (see #325).

### Steps to reproduce

1. Load the MN exp027e KR infusion study - see https://github.com/PrincetonUniversity/tracebase-rabinowitz-data/pull/4

*Note:* This should be more easily reproduced when #325 is completed.

This was discovered attempting to view PeakData for the (incorrectly loaded) "MN_exp027e_acute KR infusion" study on tracebase-dev, which caused a `NoReverseMatch` exception.

1. View study list - http://tracebase-dev.princeton.edu/DataRepo/studies/
2. View Study "MN_exp027e_acute KR infusion" - http://tracebase-dev.princeton.edu/DataRepo/studies/3/
3. Click on link to view PeakData - http://tracebase-dev.princeton.edu/DataRepo/search_basic/Study/id/exact/3/peakdata/

### Current behavior

```
Environment:


Request Method: GET
Request URL: http://tracebase-dev.princeton.edu/DataRepo/search_basic/Study/id/exact/3/peakdata/

Django Version: 3.2.5
Python Version: 3.8.8
Installed Applications:
['DataRepo.apps.DatarepoConfig',
 'django.contrib.admin',
 'django.contrib.auth',
 'django.contrib.contenttypes',
 'django.contrib.sessions',
 'django.contrib.messages',
 'django.contrib.staticfiles']
Installed Middleware:
['django.middleware.security.SecurityMiddleware',
 'django.contrib.sessions.middleware.SessionMiddleware',
 'django.middleware.common.CommonMiddleware',
 'django.middleware.csrf.CsrfViewMiddleware',
 'django.contrib.auth.middleware.AuthenticationMiddleware',
 'django.contrib.messages.middleware.MessageMiddleware',
 'django.middleware.clickjacking.XFrameOptionsMiddleware']


Template error:
In template /var/www/tracebase/DataRepo/templates/DataRepo/search/results/peakdata.html, error at line 236
   Reverse for 'compound_detail' with arguments '('',)' not found. 1 pattern(s) tried: ['DataRepo/compounds/(?P<pk>[0-9]+)/$']
   226 : 
   227 :                             <!-- Treatment -->
   228 :                             <td>
   229 :                                 {{ pd.peak_group.msrun.treatment.name }}
   230 :                             </td>
   231 : 
   232 :                             <!-- Tracer Compound -->
   233 :                             <td>
   234 :                                 <!-- Put displayed link text first for sorting -->
   235 :                                 <div style="display:none;">{{ pd.peak_group.msrun.sample.animal.tracer_compound.name }}</div>
   236 :                                 <a href=" {% url 'compound_detail' pd.peak_group.msrun.sample.animal.tracer_compound.id %} ">
   237 :                                     {{ pd.peak_group.msrun.sample.animal.tracer_compound.name }}
   238 :                                 </a>
   239 :                             </td>
   240 : 
   241 :                             <!-- Tracer Infusion Rate (ul/min/g) -->
   242 :                             <td class="text-end">
   243 :                                 {{ pd.peak_group.msrun.sample.animal.tracer_infusion_rate }}
   244 :                             </td>
   245 : 
   246 :                             <!-- Tracer Infusion Concentration (mM) -->


Traceback (most recent call last):
  File "/usr/local/tracebase/lib/python3.8/site-packages/django/core/handlers/exception.py", line 47, in inner
    response = get_response(request)
  File "/usr/local/tracebase/lib/python3.8/site-packages/django/core/handlers/base.py", line 181, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/var/www/tracebase/DataRepo/views.py", line 266, in search_basic
    return render(
  File "/usr/local/tracebase/lib/python3.8/site-packages/django/shortcuts.py", line 19, in render
    content = loader.render_to_string(template_name, context, request, using=using)
  File "/usr/local/tracebase/lib/python3.8/site-packages/django/template/loader.py", line 62, in render_to_string
    return template.render(context, request)
  File "/usr/local/tracebase/lib/python3.8/site-packages/django/template/backends/django.py", line 61, in render
    return self.template.render(context)
  File "/usr/local/tracebase/lib/python3.8/site-packages/django/template/base.py", line 170, in render
    return self._render(context)
  File "/usr/local/tracebase/lib/python3.8/site-packages/django/template/base.py", line 162, in _render
    return self.nodelist.render(context)
  File "/usr/local/tracebase/lib/python3.8/site-packages/django/template/base.py", line 938, in render
    bit = node.render_annotated(context)
  File "/usr/local/tracebase/lib/python3.8/site-packages/django/template/base.py", line 905, in render_annotated
    return self.render(context)
  File "/usr/local/tracebase/lib/python3.8/site-packages/django/template/loader_tags.py", line 150, in render
    return compiled_parent._render(context)
  File "/usr/local/tracebase/lib/python3.8/site-packages/django/template/base.py", line 162, in _render
    return self.nodelist.render(context)
  File "/usr/local/tracebase/lib/python3.8/site-packages/django/template/base.py", line 938, in render
    bit = node.render_annotated(context)
  File "/usr/local/tracebase/lib/python3.8/site-packages/django/template/base.py", line 905, in render_annotated
    return self.render(context)
  File "/usr/local/tracebase/lib/python3.8/site-packages/django/template/loader_tags.py", line 62, in render
    result = block.nodelist.render(context)
  File "/usr/local/tracebase/lib/python3.8/site-packages/django/template/base.py", line 938, in render
    bit = node.render_annotated(context)
  File "/usr/local/tracebase/lib/python3.8/site-packages/django/template/base.py", line 905, in render_annotated
    return self.render(context)
  File "/usr/local/tracebase/lib/python3.8/site-packages/django/template/loader_tags.py", line 195, in render
    return template.render(context)
  File "/usr/local/tracebase/lib/python3.8/site-packages/django/template/base.py", line 172, in render
    return self._render(context)
  File "/usr/local/tracebase/lib/python3.8/site-packages/django/template/base.py", line 162, in _render
    return self.nodelist.render(context)
  File "/usr/local/tracebase/lib/python3.8/site-packages/django/template/base.py", line 938, in render
    bit = node.render_annotated(context)
  File "/usr/local/tracebase/lib/python3.8/site-packages/django/template/base.py", line 905, in render_annotated
    return self.render(context)
  File "/usr/local/tracebase/lib/python3.8/site-packages/django/template/loader_tags.py", line 53, in render
    result = self.nodelist.render(context)
  File "/usr/local/tracebase/lib/python3.8/site-packages/django/template/base.py", line 938, in render
    bit = node.render_annotated(context)
  File "/usr/local/tracebase/lib/python3.8/site-packages/django/template/base.py", line 905, in render_annotated
    return self.render(context)
  File "/usr/local/tracebase/lib/python3.8/site-packages/django/template/defaulttags.py", line 312, in render
    return nodelist.render(context)
  File "/usr/local/tracebase/lib/python3.8/site-packages/django/template/base.py", line 938, in render
    bit = node.render_annotated(context)
  File "/usr/local/tracebase/lib/python3.8/site-packages/django/template/base.py", line 905, in render_annotated
    return self.render(context)
  File "/usr/local/tracebase/lib/python3.8/site-packages/django/template/defaulttags.py", line 312, in render
    return nodelist.render(context)
  File "/usr/local/tracebase/lib/python3.8/site-packages/django/template/base.py", line 938, in render
    bit = node.render_annotated(context)
  File "/usr/local/tracebase/lib/python3.8/site-packages/django/template/base.py", line 905, in render_annotated
    return self.render(context)
  File "/usr/local/tracebase/lib/python3.8/site-packages/django/template/defaulttags.py", line 312, in render
    return nodelist.render(context)
  File "/usr/local/tracebase/lib/python3.8/site-packages/django/template/base.py", line 938, in render
    bit = node.render_annotated(context)
  File "/usr/local/tracebase/lib/python3.8/site-packages/django/template/base.py", line 905, in render_annotated
    return self.render(context)
  File "/usr/local/tracebase/lib/python3.8/site-packages/django/template/loader_tags.py", line 195, in render
    return template.render(context)
  File "/usr/local/tracebase/lib/python3.8/site-packages/django/template/base.py", line 172, in render
    return self._render(context)
  File "/usr/local/tracebase/lib/python3.8/site-packages/django/template/base.py", line 162, in _render
    return self.nodelist.render(context)
  File "/usr/local/tracebase/lib/python3.8/site-packages/django/template/base.py", line 938, in render
    bit = node.render_annotated(context)
  File "/usr/local/tracebase/lib/python3.8/site-packages/django/template/base.py", line 905, in render_annotated
    return self.render(context)
  File "/usr/local/tracebase/lib/python3.8/site-packages/django/template/loader_tags.py", line 53, in render
    result = self.nodelist.render(context)
  File "/usr/local/tracebase/lib/python3.8/site-packages/django/template/base.py", line 938, in render
    bit = node.render_annotated(context)
  File "/usr/local/tracebase/lib/python3.8/site-packages/django/template/base.py", line 905, in render_annotated
    return self.render(context)
  File "/usr/local/tracebase/lib/python3.8/site-packages/django/template/defaulttags.py", line 211, in render
    nodelist.append(node.render_annotated(context))
  File "/usr/local/tracebase/lib/python3.8/site-packages/django/template/base.py", line 905, in render_annotated
    return self.render(context)
  File "/usr/local/tracebase/lib/python3.8/site-packages/django/template/defaulttags.py", line 211, in render
    nodelist.append(node.render_annotated(context))
  File "/usr/local/tracebase/lib/python3.8/site-packages/django/template/base.py", line 905, in render_annotated
    return self.render(context)
  File "/usr/local/tracebase/lib/python3.8/site-packages/django/template/defaulttags.py", line 312, in render
    return nodelist.render(context)
  File "/usr/local/tracebase/lib/python3.8/site-packages/django/template/base.py", line 938, in render
    bit = node.render_annotated(context)
  File "/usr/local/tracebase/lib/python3.8/site-packages/django/template/base.py", line 905, in render_annotated
    return self.render(context)
  File "/usr/local/tracebase/lib/python3.8/site-packages/django/template/defaulttags.py", line 446, in render
    url = reverse(view_name, args=args, kwargs=kwargs, current_app=current_app)
  File "/usr/local/tracebase/lib/python3.8/site-packages/django/urls/base.py", line 86, in reverse
    return resolver._reverse_with_prefix(view, prefix, *args, **kwargs)
  File "/usr/local/tracebase/lib/python3.8/site-packages/django/urls/resolvers.py", line 694, in _reverse_with_prefix
    raise NoReverseMatch(msg)

Exception Type: NoReverseMatch at /DataRepo/search_basic/Study/id/exact/3/peakdata/
Exception Value: Reverse for 'compound_detail' with arguments '('',)' not found. 1 pattern(s) tried: ['DataRepo/compounds/(?P<pk>[0-9]+)/$']
```

### Expected behavior

A list of PeakData records belonging to that Study

### Suggested Change

Something with the compound id seems incorrect for at least one row.

### Comment

Note, the corresponding tsv format should also be affected by this...

-----

### Assumptions

none

### Requirements

- Allow no tracer foreign key to exist
- Show fields that require a tracer as "None"

### Limitations

none

### Affected/Changed Components

- peakdata.html
- peakdata.tsv

I will also do:

- peakgroups.html
- peakgroups.tsv

The fcirc templates should be unaffected because they only show data that is associated with tracers.

#### GUI Change description

No exception thrown when trying to search (and download) and getting animals without tracers.

#### Code Change Description *(Pseudocode optional)*

Create a conditional that checks for a tracer foreign key before rendering any compound data for the tracer.

#### Tests

Render a record with no tracer.

### PR DESCRIPTION
**Pull request by lparsons**
## Summary Change Description

Makes `tracer_compound` and `tracer_labeled_count` non-nullable fields. This prevents other issues downstream.

## Affected Issue Numbers

- Resolves #312

## Code Review Notes

Ensure that Animals have a valid tracer_compound and
tracer_labeled_count value. This prevents other issues downstream.

Update Animal tests to account for this change

Update DataFrame tests to account for this change

The migration uses a one-time default to set any missing values to `compound_id=5` which is `arginine` most likely. It also sets the `tracer_labeled_count=6`. These were the values for the study where the missing values were discovered, but reloading of the data would ensure it is correct (which we'll do on tracebase-dev). The migration just allows one to avoid having to reload the entire db.

## Checklist

- [x] All issue requirements satisfied (or no linked issues)
- [x] [Linting passes](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#linting).
- [x] [Migrations created & committed *(or no model changes)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#migration-process)
- [x] [Tests implemented *(or no code changes)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#test-implementation)
- [x] [All tests pass](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#quality-control)

### CODE REVIEW

File: DataRepo/models.py [hunk 1]
```diff
@@ -356,27 +356,21 @@ class Animal(models.Model, TracerLabeledClass):
...
         default=TracerLabeledClass.CARBON,
-        blank=True,
         help_text="The type of atom that is labeled in the tracer compound "
         '(e.g. "C", "H", "O").',
     )
     # NOTE: encoding atom count as an integer, NOT a float, as I have seen in
     # some example files
     tracer_labeled_count = models.PositiveSmallIntegerField(
-        null=True,
-        blank=True,
```
Comment (hepcat72) [inline DataRepo/models.py:379]: NB. I feel like going with the default implies the requirement decision wasn't explicit.  Actually setting them to false would show it was an intentional decision for anyone coming back to this code, but it's up to you.

**Review by hepcat72 — APPROVED** [review]
An efficient fix to the problem. Kicks the resolution back to the user.

File: DataRepo/models.py (hunk 1 above)
Comment (lparsons) [reply DataRepo/models.py:379]: I think you mean that we should consider removing the `default=TracerLabeledClass.CARBON` from `tracer_labeled_atom`. And I agree with that. I don't see the value in the default setting.

**Comment by jcmatese** [comment]
I think we might need to run this by the full lab, because this would limit the project to only tracer-infused animals.  But perhaps that is what they wish...  Alternatively, we could roll this into the "multiple tracers?" issue, as that would pop that out of the animal model altogether.  Feel free to push this through for beta release, if it is dependent on it (as I won't get a chance to look at this is detail until Thursday).     Do you know the nature of the issue why those data were loaded sans-tracer, if it was defined in the file?

**Comment by lparsons** [comment]
I don't know how/why the data was mis-loaded. The model was set-up to allow no tracers. Im open to other suggestions, but I don't think it should be left as is since the valid data caused exceptions.

Lance Parsons
Research Software Engineer
Lewis-Sigler Institute for Integrative Genomics
Princeton University
________________________________

**Comment by lparsons** [comment]
OK, looks like we want to support animals with no tracers. I don't have any data for that just yet, but I'll probably modify #312 (or just close and create some new issues).
1. Issue to add an example study that includes animals with tracer
2. Issue to track the bug in the template rendering when an animal has no tracer

**Pull request closed without merging**